subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestClassSelectionMatcher.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.filter;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static org.apache.commons.lang.StringUtils.splitPreserveAllTokens;
+import static org.apache.commons.lang.StringUtils.substringAfterLast;
+
+/**
+ * Excludes as many classes as possible before sending them to {@link org.gradle.api.internal.tasks.testing.TestClassProcessor}
+ * in order to improve test performance. Matching rules are:
+ *
+ * <ul>
+ * <li>If the pattern starts with an upper-case letter, it's used to match simple class name.
+ * <li>Otherwise, it's used to match full qualified class name.
+ * </ul>
+ *
+ * In both cases, wildcards are supported.
+ */
+public class TestClassSelectionMatcher {
+    private final List<TestClassPattern> patterns = new ArrayList<TestClassPattern>();
+
+    public TestClassSelectionMatcher(Collection<String> includedTests, Collection<String> includedTestsCommandLine) {
+        for (String pattern : includedTests) {
+            patterns.add(TestClassPattern.fromPattern(pattern));
+        }
+        for (String pattern : includedTestsCommandLine) {
+            patterns.add(TestClassPattern.fromPattern(pattern));
+        }
+    }
+
+    public boolean mayBeIncluded(String fullQualifiedName) {
+        if (patterns.isEmpty()) {
+            return true;
+        }
+        String simpleName = getSimpleName(fullQualifiedName);
+        for (TestClassPattern pattern : patterns) {
+            if (pattern.mayBeIncluded(fullQualifiedName, simpleName)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public static String getSimpleName(String fullQualifiedName) {
+        String simpleName = substringAfterLast(fullQualifiedName, ".");
+        if ("".equals(simpleName)) {
+            return fullQualifiedName;
+        }
+        return simpleName;
+    }
+
+    private abstract static class TestClassPattern {
+        protected String[] segments;
+
+        private boolean mayBeIncluded(String fullQualifiedName, String simpleName) {
+            if (patternStartsWithUpperCase()) {
```

Comment:
❌ This should be two different classes (simple name pattern, qualified name pattern).

Reply:
Actually that's what I did in last commit. The problem is, we need to handle pattern with or without wildcard separately, which means 4 classes:

- Pattern with wildcard, handle full qualified name
- Pattern with wildcard, handle simple name
- Pattern without wildcard, handle full qualified name
- Pattern without wildcard, handle simple name

That's too many cases. Instead, I think simple name can be considered as special full qualified name, in this way we can reduce these cases to 2. That's why I made the current class design.

Reply:
You can use composition over inheritance here. If most of the code is the same and only 2 aspects need to be different, create interfaces for these two aspects and implementation classes for each. Then combine those in the overall algorithm instead of using if-statements.